solidity/contracts/vault/TBTCOptimisticMinting.sol

```diff
@@ -0,0 +1,518 @@
+//               ▐████▌    ▐████▌
+// ██████████████     ▐████▌     ██████████████
+// ██████████████     ▐████▌     ██████████████
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+
+pragma solidity 0.8.17;
+
+import "../bridge/Bridge.sol";
+import "../bridge/Deposit.sol";
+import "../GovernanceUtils.sol";
+
+/// @title TBTC Optimistic Minting
+/// @notice The Optimistic Minting mechanism allows to mint TBTC before
+///         `TBTCVault` receives the Bank balance. There are two permissioned
+///         sets in the system: Minters and Guardians, both set up in 1-of-n
+///         mode. Minters observe the revealed deposits and request minting TBTC.
+///         Any single Minter can perform this action. There is a 3 hours delay
+///         between the time of the request from a Minter to the time TBTC is
+///         minted. During the time of the delay, any Guardian can cancel the
+///         minting.
+/// @dev This functionality is a part of `TBTCVault`. It is implemented in
+///      a separate abstract contract to achieve better separation of concerns
+///      and easier-to-follow code.
+abstract contract TBTCOptimisticMinting is Ownable {
+    // Represents optimistic minting request for the given deposit revealed
+    // to the Bridge.
+    struct OptimisticMintingRequest {
+        // UNIX timestamp at which the optimistic minting was requested.
+        uint64 requestedAt;
+        // UNIX timestamp at which the optimistic minting was finalized.
+        // 0 if not yet finalized.
+        uint64 finalizedAt;
+    }
+
+    /// @notice The time delay that needs to pass between initializing and
+    ///         finalizing the upgrade of governable parameters.
+    uint256 public constant GOVERNANCE_DELAY = 24 hours;
+
+    Bridge public immutable bridge;
+
+    /// @notice Indicates if the optimistic minting has been paused. Only the
+    ///         Governance can pause optimistic minting. Note that the pause of
+    ///         the optimistic minting does not stop the standard minting flow
+    ///         where wallets sweep deposits.
+    bool public isOptimisticMintingPaused;
+
+    /// @notice Divisor used to compute the treasury fee taken from each
+    ///         optimistically minted deposit and transferred to the treasury
+    ///         upon finalization of the optimistic mint. This fee is computed
+    ///         as follows: `fee = amount / optimisticMintingFeeDivisor`.
+    ///         For example, if the fee needs to be 2% of each deposit,
+    ///         the `optimisticMintingFeeDivisor` should be set to `50` because
+    ///         `1/50 = 0.02 = 2%`.
+    ///         Note that the optimistic minting fee does not replace the
+    ///         deposit treasury fee cut by the Bridge. Both fees are deducted
+    ///         from the minted token amount.
```

Comment:
We could apply the optimistic minting fee based on the original deposit amount, but that would lead to inconsistent amounts if the optimistic minting fee would be high. It is easiest to explain it with an example. Say that the deposit amount is 1000 and that the Bridge fee is 20. 980 is what lands in TBTCVault. If the optimistic deposit fee was anything > 98%, we would be printing tokens because only 980 out of the original 1000 lands in the vault. In other words, the vault can cut a fee only from what it is actually receiving. We could probably introduce some kind of validation between fees but that would break the separation between vaults (applications) and the Bridge.

Please see 22e18b7c1efb262eeda1be7ead8e865c69b8b14c for documentation improvements.

Reply:
Agreed with both of y'all, let's do it in the next version.